File: transports/record.go
package transports

import (
	"tempo/runtime"
)

type SendValue struct {
	Value     any
	Receivers []string
}

type RecvValue struct {
	Value  any
	Sender string
}

type Recorder struct {
	inner    runtime.Transport
	sends    []SendValue
	receives []RecvValue
}

func NewRecorder(transport runtime.Transport) *Recorder {
	return &Recorder{
		inner:    transport,
		sends:    []SendValue{},
		receives: []RecvValue{},
	}
}

func (r *Recorder) SendValues() []SendValue {
	return r.sends
}

func (r *Recorder) ReceivedValues() []RecvValue {
	result := []RecvValue{}
	for _, val := range r.receives {
		async := val.Value.(*runtime.Async[any])
		result = append(result, RecvValue{
			Value:  runtime.GetAsync(async),
			Sender: val.Sender,
		})
	}
	return result
}

// Recv implements runtime.Transport.
func (r *Recorder) Recv(role string, value any) *runtime.Async[any] {
	result := r.inner.Recv(role, value)

	r.receives = append(r.receives, RecvValue{
		Value:  result,
		Sender: role,
	})

	return result
}

// Send implements runtime.Transport.
func (r *Recorder) Send(value any, roles ...string) {
	r.inner.Send(value, roles...)

	r.sends = append(r.sends, SendValue{
		Value:     value,
		Receivers: roles,
	})
}

File: typescript/transports/record.ts
import { Transport } from "../runtime.ts";

/**
 * A value that has been sent to a set of receivers.
 */
export type SendValue = {
  value: unknown;
  receivers: string[];
};

/**
 * A value received by a given sender.
 */
export type RecvValue = {
  value: unknown;
  sender: string;
};

/**
 * Recoder wraps another {@link Transport} implementation and records every message sent and received.
 */
export class Recorder implements Transport {
  #inner: Transport;
  #sends: SendValue[];
  #receives: RecvValue[];

  constructor(transport: Transport) {
    this.#inner = transport;
    this.#sends = [];
    this.#receives = [];
  }

  get sends(): SendValue[] {
    return this.#sends;
  }

  get receives(): RecvValue[] {
    return this.#receives;
  }

  send<T>(value: T, ...roles: string[]): void {
    this.#sends.push({
      receivers: roles,
      value,
    });

    this.#inner.send(value, ...roles);
  }

  async recv<T>(role: string): Promise<T> {
    const value: T = await this.#inner.recv(role);

    this.#receives.push({
      sender: role,
      value,
    });

    return value;
  }
}
